hypothesis/lms
# Return a redirect to automatically log users in to Vitalsource

For:

 * https://github.com/orgs/hypothesis/projects/71/views/1

## What's changed?

Previous to this PR we would direct the user straight to the bookshelf on the correct book. This required the user to have independently signed into the bookshelf for this to work.

Now we:

 * Check if the user has a licences
    *  If not show an error
 * Generate a redirect link that signs the user in, and then redirects to the shelf

This means the user does not always need to log in

### What's not so great about this?

If the user does not have a licence then we show them an error instead of showing them the book viewer. However users can sometimes be automatically granted a licence by the book viewer. Unfortunately the redirect mechanism doesn't take this into account and fails.

This means we are deciding between our error page and the one generated by the redirect. It was decided that our error page could contain more helpful information about what had gone wrong than the default. At the moment there's no very good way to help the user to resolve the situation.

## ~Do not merge!~

~We can't deploy this until existing customers have settings added to their application instances.~

We've merged in this PR now which should make this safe to release:

 * https://github.com/hypothesis/lms/pull/4351

## Testing notes

Checking SSO:

 * Visit: https://bookshelf.vitalsource.com/ and ensure you are not logged in
 * Start all the things
 * Make sure you've run `make devdata` recently in LMS
 * Visit: https://hypothesis.instructure.com/courses/125/assignments/3153 as the teacher account
 * You should be redirected to the book eventually (it's a little slow for me)
 * Visit: https://bookshelf.vitalsource.com
 * You will see you are now logged in

When there's no licence:

 * Visit: https://bookshelf.vitalsource.com/ and ensure you are not logged in
 * Visit: https://hypothesis.instructure.com/courses/125/assignments/3127
 * You should see an unhandled error "Sorry, but something went wrong."

When there's no user (faking it):

 * Edit `lms.services.vitalsource._client.py:get_user_credentials()`
 * Comment out the if above the raised error
 * Visit: https://hypothesis.instructure.com/courses/125/assignments/3153
 * You should see an unhandled error "Sorry, but something went wrong."
